services/web/server/src/simcore_service_webserver/users/_users_repository.py

```diff
@@ -0,0 +1,422 @@
+            users_pre_registration_details.c.extras,
+            users.c.status,
+            invited_by,
+        )
+
+        left_outer_join = (
+            sa.select(*columns)
+            .select_from(
+                users_pre_registration_details.outerjoin(
+                    users, users.c.id == users_pre_registration_details.c.user_id
+                )
+            )
+            .where(users_pre_registration_details.c.pre_email.like(email_like))
+        )
+        right_outer_join = (
+            sa.select(*columns)
+            .select_from(
+                users.outerjoin(
+                    users_pre_registration_details,
+                    users.c.id == users_pre_registration_details.c.user_id,
+                )
+            )
+            .where(users.c.email.like(email_like))
+        )
+
+        result = await conn.stream(sa.union(left_outer_join, right_outer_join))
+        return [row async for row in result]
+
+
+async def get_user_products(
+    engine: AsyncEngine,
+    connection: AsyncConnection | None = None,
+    *,
+    user_id: UserID,
+) -> list[Row]:
+    async with pass_or_acquire_connection(engine, connection) as conn:
+        product_name_subq = (
+            sa.select(products.c.name)
+            .where(products.c.group_id == groups.c.gid)
+            .label("product_name")
+        )
+        products_gis_subq = sa.select(products.c.group_id).distinct().subquery()
+        query = (
+            sa.select(
+                groups.c.gid,
+                product_name_subq,
+            )
+            .select_from(
+                users.join(user_to_groups, user_to_groups.c.uid == users.c.id).join(
+                    groups,
+                    (groups.c.gid == user_to_groups.c.gid)
+                    & groups.c.gid.in_(products_gis_subq),
+                )
+            )
+            .where(users.c.id == user_id)
+            .order_by(groups.c.gid)
+        )
+        result = await conn.stream(query)
+        return [row async for row in result]
+
+
+async def new_user_details(
+    engine: AsyncEngine,
+    connection: AsyncConnection | None = None,
+    *,
+    email: str,
+    created_by: UserID,
+    **other_values,
+) -> None:
+    async with transaction_context(engine, connection) as conn:
+        await conn.execute(
+            sa.insert(users_pre_registration_details).values(
+                created_by=created_by, pre_email=email, **other_values
+            )
+        )
+
+
+async def get_user_billing_details(
+    engine: AsyncEngine, connection: AsyncConnection | None = None, *, user_id: UserID
+) -> UserBillingDetails:
+    """
+    Raises:
+        BillingDetailsNotFoundError
+    """
+    async with pass_or_acquire_connection(engine, connection) as conn:
+        query = UsersRepo.get_billing_details_query(user_id=user_id)
+        result = await conn.stream(query)
+        row = await result.fetchone()
+        if not row:
```

Comment:
why do you use the stream here? last time we checked, it's anyway all in memory. and there is no sync vs async issue.

Reply:
yes